init.ps1

```diff
@@ -65,6 +65,19 @@ if (!$NoPrerequisites) {
     if ($env:OS -eq 'Windows_NT') {
         $EnvVars['PROCDUMP_PATH'] = & "$PSScriptRoot\azure-pipelines\Get-ProcDump.ps1"
     }
+
+    # Install specific macOS workload version in order to support usage on older .NET and macOS
+    $rollbackFile = Join-Path ([System.IO.Path]::GetTempPath()) ([System.Guid]::NewGuid())
+    @{
+        "microsoft.net.sdk.macos" = "12.3.2372/7.0.100"
```

Comment:
Can you explain to me more about this requirement? Does this mean that `dotnet workload restore` is not enough?
I wonder too if this is somehow related to msbuild's failing import error.

Reply:
The macOS workload is necessary in order to target `net6.0-macos`. That target framework is not available without the workload being installed.

The .NET SDK makes it difficult to specify a preferred version of a given workload. You have to use a rollback file to achieve this. Without specifying the version, the SDK will pull the latest for the SDK version specified in `global.json` (without taking target framework into account). So, even though this project is `net6.0-macos`, it would be referencing a newer `Microsoft.macOS.dll` that is not compatible with projects using the .NET 6 SDK and an older macos workload. To fix this, we specify an older workload version that is compatible with an older .NET/workload.

>  I wonder too if this is somehow related to msbuild's failing import error.

Yes. `dotnet workload restore` appears to be broken, at least with the macOS workload. It's like it tries to load targets before they have been installed. Tomas filed a bug about it recently: https://github.com/dotnet/sdk/issues/30230. It would be great if the .NET SDK team fixed that. We can't depend on `dotnet restore` for CI, because it would pick up the wrong workload version as described earlier, but for local development it would be very convenient.

Reply:
Thanks. That's helpful, and I've upvoted that issue.
How do dotnet workloads work when the SDK to be installed into is in an elevated location like Program Files?
It's very odd to me that init.ps1 works, but running this command outside of init fails, where `workloads.json` refers to the json file your init.ps1 creates:

```
> dotnet workload install macos --from-rollback-file workloads.json

Downloading Microsoft.macOS.Sdk.Msi.x64 (13.1.1024)
Workload installation failed. Rolling back installed packs...
Workload installation failed: One or more errors occurred. (microsoft.macos.sdk.msi.x64::13.1.1024 is not found in NuGet feeds https://api.nuget.org/v3/index.json;https://pkgs.dev.azure.com/dnceng/public/_packaging/test-tools/nuget/v3/index.json".)
```

Reply:
Notice it's picking 13.1.1024, which is a different version. It's newer, and I don't see it in nuget.

If you have multiple SDK versions installed, and you're running `workload install` from outside of a context where a `global.json` is specifying your SDK version, you can get different results based on whatever the default SDK is.

>  How do dotnet workloads work when the SDK to be installed into is in an elevated location like Program Files?

On Mac, when installing workloads to the machine-wide .NET SDK installation (`/usr/local/share/dotnet`), you have to use `sudo` to elevate. Presumably something similar is needed on Windows?

Reply:
I do have the 7.0.200 installed, as of this morning when a VS update installed it. But all my dotnet workload install attempts were within the directory containing global.json, so evidently my repo's 7.0.101 constraint wasn't effective at retaining behavior from that SDK in this regard.

Windows doesn't need `sudo` to elevate. It prompts with interactive UI as needed. It turns out that there are a collection of bugs in the SDK around this area that explains the various failures. :(